library/src/cache/signing.rs

```diff
@@ -0,0 +1,133 @@
+use anyhow::{bail, Context, Result};
+use base64::Engine;
+use ring::signature;
+use std::path::Path;
+// https://stackoverflow.com/questions/67087597/is-it-possible-to-use-rusts-log-info-for-tests
+#[cfg(test)]
+use std::{println as info, println as debug}; // Workaround to use println! for logs.
+
+/// Reads the file at `path` and returns the SHA-256 hash of its contents as a String.
+pub fn hash_file<P: AsRef<Path>>(path: P) -> Result<String> {
+    use sha2::{Digest, Sha256}; // `Digest` is needed for `Sha256::new()`;
+
+    let mut file = std::fs::File::open(path)?;
+    let mut hasher = Sha256::new();
+    std::io::copy(&mut file, &mut hasher)?;
+    let hash = hasher.finalize();
+    Ok(hex::encode(hash))
+}
+
+/// `public_key` is a DER base64-encoded RSA public key.
+///   openssl rsa -pubin \
+///     -in public_key.pem \
+///     -inform PEM \
+///     -RSAPublicKey_out \
+///     -outform DER \
+///     -out public_key.der
+pub fn check_signature(message: &str, signature: &str, public_key: &str) -> Result<()> {
+    debug!("Message is {}", message);
+    debug!("Public key is {:?}", public_key);
+    debug!("Signature is {}", signature);
+
+    // public.der
+    let public_key_bytes = base64::prelude::BASE64_STANDARD
+        .decode(public_key)
+        .with_context(|| format!("Failed to decode public_key: {}", public_key))?;
+
+    let public_key =
+        signature::UnparsedPublicKey::new(&signature::RSA_PKCS1_2048_8192_SHA256, public_key_bytes);
+    let decoded_sig = match base64::prelude::BASE64_STANDARD.decode(signature) {
+        Ok(sig) => sig,
+        Err(e) => {
+            bail!("Failed to decode signature: {:?}", e);
+        }
+    };
```

Comment:
Is this just .context()?

Reply:
Or .map_err()?

Reply:
Possibly could be, let me see if there's a way to include the error message in a `.context`

Reply:
`map_err` works, just had to use `anyhow`'s Error type, which is what is "thrown" by `bail!`:

```rust
    let decoded_sig = base64::prelude::BASE64_STANDARD
        .decode(signature)
        .map_err(|e| anyhow::Error::msg(format!("Failed to decode signature: {:?}", e)))?;
```